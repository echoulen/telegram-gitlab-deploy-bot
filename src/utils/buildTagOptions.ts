import {List} from "immutable";
import {Option} from "../types/Option";

export enum TagType {
  RC,
  PRODUCTION,
}

export function buildTagOptions(lastTag: {name: string}, type: TagType): List<Option> {
  if (!lastTag) {
    switch (type) {
      case TagType.RC:
        return List([
          {text: `0.0.1-rc1`, callback_data: `add-tag-v0.0.1-rc1`},
          {text: `0.1.0-rc1`, callback_data: `add-tag-v0.1.0-rc1`},
          {text: `1.0.0-rc1`, callback_data: `add-tag-v1.0.0-rc1`},
          {text: "取消", callback_data: `cancel-tag`},
        ]);
      default:
        return List([
          {text: `0.0.1`, callback_data: `add-tag-v0.0.1`},
          {text: `0.1.0`, callback_data: `add-tag-v0.1.0`},
          {text: `1.0.0`, callback_data: `add-tag-v1.0.0`},
          {text: "取消", callback_data: `cancel-tag`},
        ]);
    }
  }

  const vs = lastTag.name.split("-");
  const version = vs[0];
  const versions = version.split(".");
  const main = Number(versions[0].replace("v", ""));
  const minor = Number(versions[1]);
  const ver = Number(versions[2]);

  if (vs[1] && vs[1].indexOf("rc") === 0) {
    const rcNo = vs[1].slice(2, vs[1].length);

    switch (type) {
      case TagType.RC:
        return List([
          {
            text: `${main}.${minor}.${ver}-rc${Number(rcNo) + 1}`,
            callback_data: `add-tag-v${main}.${minor}.${ver}-rc${Number(rcNo) + 1}`,
          },
          {text: "取消", callback_data: `cancel-tag`},
        ]);
      default:
        return List([
          {
            text: `${main}.${minor}.${ver}`,
            callback_data: `add-tag-v${main}.${minor}.${ver}`,
          },
          {text: "取消", callback_data: `cancel-tag`},
        ]);
    }

  } else {
    switch (type) {
      case TagType.RC:
        return List([
          {
            text: `${main}.${minor}.${ver + 1}-rc1`,
            callback_data: `add-tag-v${main}.${minor}.${ver + 1}-rc1`,
          },
          {
            text: `${main}.${minor + 1}.0-rc1`,
            callback_data: `add-tag-v${main}.${minor + 1}.0-rc1`,
          },
          {
            text: `${main + 1}.0.0-rc1`,
            callback_data: `add-tag-v${main + 1}.0.0-rc1`,
          },
          {
            text: "取消",
            callback_data: `cancel-tag`,
          }
        ]);
      default:
        return List([
          {
            text: `${main}.${minor}.${ver + 1}`,
            callback_data: `add-tag-v${main}.${minor}.${ver + 1}`,
          },
          {
            text: `${main}.${minor + 1}.0`,
            callback_data: `add-tag-v${main}.${minor + 1}.0`,
          },
          {
            text: `${main + 1}.0.0`,
            callback_data: `add-tag-v${main + 1}.0.0`,
          },
          {
            text: "取消",
            callback_data: `cancel-tag`,
          }
        ]);
    }
  }
}
