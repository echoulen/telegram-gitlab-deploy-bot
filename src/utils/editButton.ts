export function editButton(bot, query, inlineKeyboards): void {
  bot.editMessageReplyMarkup(
    {inline_keyboard: [inlineKeyboards]},
    {
      chat_id: query.message.chat.id,
      message_id: query.message.message_id
    },
  );
}
