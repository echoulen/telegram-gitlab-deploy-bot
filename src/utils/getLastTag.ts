import {List} from "immutable";

export function getLastTag(tags: Array<{name: string}>): {name: string} {
  return List(tags)
    .map(({name}) => {
      const v = name.replace("v", "").split("-rc");
      const mainVersion = List(v[0].split("."))
        .map(it => it.length === 1 ? `0${it}` : it)
        .join("");
      const rcVersion = v[1] ? v[1].length === 1 ? `0${v[1]}` : v[1] : "00";
      const value = mainVersion + rcVersion;
      return {name, value: Number(value)};
    })
    .filter(it => it.value > 0)
    .sortBy(it => it.value)
    .last();
}
