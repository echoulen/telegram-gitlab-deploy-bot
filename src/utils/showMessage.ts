export const showMessage = (bot, query, message: string) => {
  bot.editMessageText(message, {
    chat_id: query.message.chat.id,
    message_id: query.message.message_id
  });
  bot.answerCallbackQuery(query.id, {
    text: message,
    cache_time: 3
  });
};
