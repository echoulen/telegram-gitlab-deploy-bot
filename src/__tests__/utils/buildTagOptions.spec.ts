import {buildTagOptions} from "../../utils/buildTagOptions";

describe("buildTagOptions spec", () => {
  it("should buildTagOptions proper if tag = v0.0.1", () => {
    const options = buildTagOptions({name: "v0.0.1"});
    expect(options).toMatchSnapshot();
  });

  it("should buildTagOptions proper if tag = v2.3.5", () => {
    const options = buildTagOptions({name: "v2.3.5"});
    expect(options).toMatchSnapshot();
  });

  it("should buildTagOptions proper if tag = v2.1.0-rc5", () => {
    const options = buildTagOptions({name: "v2.1.0-rc5"});
    expect(options).toMatchSnapshot();
  });

  it("should buildTagOptions proper if tag = v1.0.0-rc10", () => {
    const options = buildTagOptions({name: "v1.0.0-rc10"});
    expect(options).toMatchSnapshot();
  });

  it("should buildTagOptions proper if tag = null", () => {
    const options = buildTagOptions(null);
    expect(options).toMatchSnapshot();
  });
});
