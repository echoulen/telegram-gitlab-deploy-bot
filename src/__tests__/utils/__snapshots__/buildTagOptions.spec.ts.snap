// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTagOptions spec should buildTagOptions proper if tag = null 1`] = `
Immutable.List [
  Object {
    "callback_data": "add-tag-v0.0.1",
    "text": "0.0.1",
  },
  Object {
    "callback_data": "add-tag-v0.1.0",
    "text": "0.1.0",
  },
  Object {
    "callback_data": "add-tag-v1.0.0",
    "text": "1.0.0",
  },
  Object {
    "callback_data": "cancel-tag",
    "text": "取消",
  },
]
`;

exports[`buildTagOptions spec should buildTagOptions proper if tag = v0.0.1 1`] = `
Immutable.List [
  Object {
    "callback_data": "add-tag-v0.0.2-rc1",
    "text": "0.0.2-rc1",
  },
  Object {
    "callback_data": "add-tag-v0.1.0-rc1",
    "text": "0.1.0-rc1",
  },
  Object {
    "callback_data": "add-tag-v1.0.0-rc1",
    "text": "1.0.0-rc1",
  },
  Object {
    "callback_data": "cancel-tag",
    "text": "取消",
  },
]
`;

exports[`buildTagOptions spec should buildTagOptions proper if tag = v1.0.0-rc10 1`] = `
Immutable.List [
  Object {
    "callback_data": "add-tag-v1.0.0",
    "text": "1.0.0",
  },
  Object {
    "callback_data": "cancel-tag",
    "text": "取消",
  },
]
`;

exports[`buildTagOptions spec should buildTagOptions proper if tag = v2.1.0-rc5 1`] = `
Immutable.List [
  Object {
    "callback_data": "add-tag-v2.1.0-rc6",
    "text": "2.1.0-rc6",
  },
  Object {
    "callback_data": "cancel-tag",
    "text": "取消",
  },
]
`;

exports[`buildTagOptions spec should buildTagOptions proper if tag = v2.3.5 1`] = `
Immutable.List [
  Object {
    "callback_data": "add-tag-v2.3.6-rc1",
    "text": "2.3.6-rc1",
  },
  Object {
    "callback_data": "add-tag-v2.4.0-rc1",
    "text": "2.4.0-rc1",
  },
  Object {
    "callback_data": "add-tag-v3.0.0-rc1",
    "text": "3.0.0-rc1",
  },
  Object {
    "callback_data": "cancel-tag",
    "text": "取消",
  },
]
`;
